File: rotateArraybyK.py
# TC: O(N) since we are iterating over each element.
# SC: O(1) since we do not use any extra space. 
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:

        if k > len(nums): return nums
        """
        Do not return anything, modify nums in-place instead.
        """
        self.reverse(nums, 0, len(nums) - 1)
        self.reverse(nums, 0, k - 1)
        self.reverse(nums, k, len(nums) - 1)

    def reverse(self, nums, left, right):
        while left < right:
            self.swap(nums, left, right)
            left += 1
            right -= 1

    def swap(self, nums, left, right):
        temp = nums[left]
        nums[left] = nums[right]
        nums[right] = temp

File: RotateArraybyK.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Array
{
    class RotateArraybyKLC
    {
        //TC: O(n)
        //SC: O(1)
        public void Rotate(int[] nums, int k)
        {
            if (nums == null || nums.Length == 0) return;
            int n = nums.Length;
            if (k > n) { k = k % n; }
            reverse(nums, 0, n - 1);
            reverse(nums, 0, k - 1);
            reverse(nums, k, n - 1);
        }
        private void reverse(int[] nums, int l, int r)
        {
            while (l < r)
            {
                swap(nums, l, r);
                l++;
                r--;
            }
        }
        private void swap(int[] nums, int l, int r)
        {
            int temp = nums[l];
            nums[l] = nums[r];
            nums[r] = temp;
        }
    }
}
